### PR TITLE
ProTip Advice and Tutoring System [DNM]

### LINKED ISSUE #5316: ProTip Advice and Tutoring System B: 130
Update 13/11/22 - Made some changes per Max's recommendations.

## Introduction and Overview
The goal of the ProTip system is to provide new players to Unitystation responsive, automatic advice that will help them navigate our complicated game. With a flexible and easy-to-use trigger system, the ProTip system could potentially become a simple and low-maintenance means of easing the new player experience. The system is largely inspired by the Learning Helper from Rimworld. This issue very much ties in with Issue #3728 on implementing Mentors. Ideally, the system should be implemented so that non-technical contributors can write and configure new Advice. This issue has/will have a bounty of USD130.

## The GUI
The GUI should be a text box that appears in the top right of the screen, approximately taking up the same space as admin tools currently do, given that the overlap between admins and players needing advice should be approximately zero. Advice is mostly just text, but it should ideally be able to contain hyperlinks to Wiki articles on more advanced topics. Control advice should also be responsive to customized keybindings. Every Advice window must contain buttons to contact a Mentor for further questions, a button linking to the Wiki front page, an X button to close the window as well as a button to adjust Advice Frequency, see below.

When the round first starts for a new player to the server, or when a player presses the “Adjust Advice frequency”-button, the hint window should pop up in the top right, offering to tailor the level of advice to the experience level the player chooses by pressing one of the following buttons:    

1. New to Space Station 13 (All Advice)
2. New to Unitystation (Only Advice on new features in Unitystation)
3. Somewhat Experienced (Life critical Advice only)
4. Robust (No Advice)

Each piece of advice will only be shown once to a given player, and then a Help section of the Escape menu will offer a list of all advice, for players to re-read if they need to be reminded of something.

## The Triggers
Advice pop-ups should trigger based on a variety of circumstances around the player. The system should be able to detect the following:

- The player holding a specific object
- The player holding one of a class of objects (Holding any gun, for example)
- Being near a world object; contains both the object ID and trigger range (The clone scanner, for example)
- Opening a GUI (Opening the hacking screen should trigger hacking advice, for example)
- Health states (The player taking damage, going unconscious, or dying)
- The player being in a hazardous area (Extreme cold or low O2 atmosphere, for example)
- Time from player joining the server
- (Cancelled) Player actions (For walking through sequential steps, like building a wall following a girder) 
- (Cancelled) Nearby players’ health states; contains both the relevant health state and trigger range (Finding someone unconscious or dead, for example)
- Changing intents or using resist, rest, and other player intents/stances
- Switching states on particular objects (Like flicking on a Stun Baton)

If multiple triggers are fired before the current protip disappears, then the ProTip’s lifespan could be significantly reduced and the new protip takes its place, but only if the advice level is of a higher priority - else, the less (or equally) important protips could be shelved and instead displayed at the next trigger instance. This would help to reduce a sudden rush of popups at the start of the game and distribute them a little more evenly throughout the round, reducing information overload and the non-ending stream of popups.

## The Anatomy of an Advice

- Any given instance of Advice must contain the following data:
- Experience level of the advice (Integer; if the player’s chosen experience level is higher than the level of the advice, the advice will not trigger)
- Trigger conditions (See above chapter)
- Text Advice (A string supporting hyperlinks)
- If possible: Reference objects or UI (Information on objects in the game world or UI elements that the Advice GUI should point to and highlight while the advice is on screen)

Ideally, all the Advices and associated triggers of a given server should be able to be contained in a simple configuration text file, where you for Each Advice have the Advice text itself, alongside invoking a trigger type and giving the relevant arguments. Ideally, the system should also support AND or OR arguments, to allow for combination and differentiation of triggers.

## The Writing
Advice should be OOC, as being encouraging and helpful to the new player takes priority over worldbuilding. The writing will be a balance between matter-of-fact brevity, and having enough meta-text and encouragement to make sure new players do not feel stressed, condescended or discouraged from continuing their discovery of the game. Additionally, the system should introduce basic systems and brief overviews of deeper systems, but not rob players of the aforementioned discovery of the game’s depth.

## The Checklist (prioritised order)

- [x] Issue #3728 should be completed before this Issue.
- [x] Implementing the Advice object, with relevant properties (See Anatomy of an Advice), in a format that non-technical contributors can work with to add new advice without needing developer backing.
- [x] Implementing the Help sub-menu in the Escape Menu, containing a button that links to the Wiki, a Mentorhelp button, and an archive of every Advice.
- [x] Implementing the Trigger system, per the relevant chapter.
- [x] Designing (possibly with assistance from the UX Team) and implementing the Advice GUI as described above, also being mindful of the GUI Style Guide.
- [ ] Implementing the ability for the Advice GUI to highlight/point to game objects and UI elements. (Low priority. Pursue only if you have a quick solution.)
- [ ] Leave a hook for Advice to be translated, when official translation support releases.

### PR DESCRIPTION
**Pull request by dapocalypse**
### Purpose
Fixes #5316 
This PR will add a ProTip Advice system.

### Notes:
Don't merge, will add stuff

### CODE REVIEW

**Comment by PerfectTangent** [comment]
what? really? cool!

File: UnityProject/Assets/Scripts/UI/Systems/Advice/AdviceController.cs [hunk 1]
```diff
@@ -22,22 +23,35 @@ public class Advice
 {
 
 	private int experienceLevel;
-	private TriggerCondition tCondition;
-	private string Text;
 
-	public Advice(int xpLevel, TriggerCondition trigger, string t)
+	/*
+	 * 1. The integer 0 on the experience level shall represent a total newbie to SS13.
+	 * 2. The integer 1 on the experience level shall represent a newbie to UnityStation.
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/UI/Systems/Advice/AdviceController.cs]: Any reason not to use enums?

Comment (dapocalypse) [reply UnityProject/Assets/Scripts/UI/Systems/Advice/AdviceController.cs]: added them

File: UnityProject/Assets/Scripts/UI/Systems/IngameMenu/GUI_IngameMenu.cs [hunk 2]
```diff
@@ -141,6 +154,16 @@ public void InitiateRestartVote()
 		CloseMenuPanel();
 	}
 
+	public void GoToWiki()
+	{
+		Application.OpenURL("https://unitystation.github.io/unitystation-wiki/");
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/UI/Systems/IngameMenu/GUI_IngameMenu.cs:159]: it'd be best if this wasn't hardcoded. Can you get a link from ``GameConfig.json`` instead?

**Comment by ThatDan123** [comment]
#6010 edits the escape menu, when you pull to update your branch dont discard your changes, override the upstream changes and then fix the menu close button and escape button values in the editor.

<details>

image
On MainIngameMenu

image
On CloseButton

</details>

**Comment by PerfectTangent** [comment]
@dapocalypse How are things going? Are you still working on this?

**Comment by dapocalypse** [comment]
I will continue working on this after SAT tommorow

**Comment by dapocalypse** [comment]
just letting you guys know that I am alive and will work on this, I have some school stuff at the moment

**Comment by vlandemart** [comment]
Closing Abandoned PR.
In case anyone resumes work on this - just reopen it.

**Pull request closed without merging**